File: app/src/main/java/no/nav/tilbakemeldingsmottak/rest/serviceklage/service/KlassifiserServiceklageService.java
package no.nav.tilbakemeldingsmottak.rest.serviceklage.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.tilbakemeldingsmottak.bigquery.serviceklager.ServiceklageEventTypeEnum;
import no.nav.tilbakemeldingsmottak.bigquery.serviceklager.ServiceklagerBigQuery;
import no.nav.tilbakemeldingsmottak.consumer.oppgave.OppgaveConsumer;
import no.nav.tilbakemeldingsmottak.consumer.oppgave.domain.EndreOppgaveRequestTo;
import no.nav.tilbakemeldingsmottak.consumer.oppgave.domain.HentOppgaveResponseTo;
import no.nav.tilbakemeldingsmottak.consumer.oppgave.domain.OpprettOppgaveRequestTo;
import no.nav.tilbakemeldingsmottak.consumer.oppgave.domain.OpprettOppgaveResponseTo;
import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException;
import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorNotFoundException;
import no.nav.tilbakemeldingsmottak.exceptions.ErrorCode;
import no.nav.tilbakemeldingsmottak.exceptions.ServerErrorException;
import no.nav.tilbakemeldingsmottak.model.Answer;
import no.nav.tilbakemeldingsmottak.model.HentSkjemaResponse;
import no.nav.tilbakemeldingsmottak.model.KlassifiserServiceklageRequest;
import no.nav.tilbakemeldingsmottak.model.Question;
import no.nav.tilbakemeldingsmottak.repository.ServiceklageRepository;
import no.nav.tilbakemeldingsmottak.rest.common.pdf.PdfService;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support.EndreOppgaveRequestToMapper;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support.OpprettOppgaveRequestToMapper;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support.ServiceklageMailHelper;
import no.nav.tilbakemeldingsmottak.serviceklage.Serviceklage;
import no.nav.tilbakemeldingsmottak.util.OidcUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import static no.nav.tilbakemeldingsmottak.config.Constants.AZURE_ISSUER;
import static no.nav.tilbakemeldingsmottak.serviceklage.ServiceklageConstants.NONE;
import static no.nav.tilbakemeldingsmottak.util.SkjemaUtils.getQuestionById;
import static org.apache.commons.lang3.StringUtils.isBlank;
import static org.apache.commons.lang3.StringUtils.isNumeric;

@Service
@Slf4j
@RequiredArgsConstructor
public class KlassifiserServiceklageService {

    private static final String KOMMUNAL_KLAGE = "Nei, serviceklagen gjelder kommunale tjenester eller ytelser";
    private static final String FORVALTNINGSKLAGE = "Nei, en forvaltningsklage";
    private static final String SKRIV_TIL_OSS = "Nei, Skriv til oss";
    private static final String JA = "Ja";
    private static final String ANNET = "Annet";
    private final ServiceklageRepository serviceklageRepository;
    private final OppgaveConsumer oppgaveConsumer;
    private final EndreOppgaveRequestToMapper endreOppgaveRequestToMapper;
    private final HentSkjemaService hentSkjemaService;
    private final PdfService pdfService;
    private final ServiceklageMailHelper mailHelper;
    private final OidcUtils oicdUtils;
    private final OpprettOppgaveRequestToMapper opprettOppgaveRequestToMapper;
    private final ServiceklagerBigQuery serviceklagerBigQuery;
    @Value("${email_serviceklage_address}")
    private String toAddress;
    @Value("${email_from_address}")
    private String fromAddress;

    public void klassifiserServiceklage(KlassifiserServiceklageRequest request, HentOppgaveResponseTo hentOppgaveResponseTo) {
        if (KOMMUNAL_KLAGE.equals(request.getBEHANDLESSOMSERVICEKLAGE())) {
            log.info("Klagen har blitt markert som en kommunal klage. Oppretter oppgave om sletting av dokument.");
            opprettSlettingOppgave(hentOppgaveResponseTo);
        }

        Serviceklage serviceklage = getOrCreateServiceklage(hentOppgaveResponseTo.getJournalpostId());
        updateServiceklage(serviceklage, request);
        serviceklageRepository.save(serviceklage);
        serviceklagerBigQuery.insertServiceklage(serviceklage, ServiceklageEventTypeEnum.KLASSIFISER_SERVICEKLAGE);

        log.info("Serviceklage med serviceklageId={} er klassifisert som {}", serviceklage.getServiceklageId(), serviceklage.getTema());

        log.info("Ferdigstille oppgave med oppgaveId={} og versjonsnummer={}", hentOppgaveResponseTo.getId(), hentOppgaveResponseTo.getVersjon());
        ferdigstillOppgave(hentOppgaveResponseTo);
        log.info("Ferdigstilt oppgave med oppgaveId={}", hentOppgaveResponseTo.getId());

        if (JA.equals(request.getKVITTERING())) {
            try {
                sendKvittering(serviceklage, hentOppgaveResponseTo);
            } catch (Exception e) {
                log.warn("Kunne ikke produsere kvittering på mail", e);
            }
        }
    }

    private void sendKvittering(Serviceklage serviceklage, HentOppgaveResponseTo hentOppgaveResponseTo) throws JsonProcessingException {

        String email = oicdUtils.getEmailForIssuer(AZURE_ISSUER).orElseThrow(() -> new ClientErrorNotFoundException("Fant ikke email-adresse i token", ErrorCode.TOKEN_EMAIL_MISSING));
        log.info("Kvittering på innsendt klassifiseringsskjema sendes til epost: {}", email);

        LinkedHashMap<String, String> questionAnswerMap = createQuestionAnswerMap(serviceklage, hentOppgaveResponseTo);

        byte[] pdf = pdfService.opprettKlassifiseringPdf(questionAnswerMap);

        mailHelper.sendEmail(fromAddress,
                email,
                "Kvittering på innsendt klassifiseringsskjema",
                "Serviceklage med oppgave-id " + hentOppgaveResponseTo.getId() + " har blitt klassifisert. " +
                        "Innholdet i ditt utfylte skjema ligger vedlagt.",
                pdf);
        log.info("Kvittering sendt på mail til saksbehandler");
    }

    private LinkedHashMap<String, String> createQuestionAnswerMap(Serviceklage serviceklage, HentOppgaveResponseTo hentOppgaveResponseTo) throws JsonProcessingException {
        LinkedHashMap<String, String> questionAnswerMap = new LinkedHashMap<>();
        HentSkjemaResponse skjemaResponse = hentOppgaveResponseTo.getJournalpostId() != null ?
                hentSkjemaService.hentSkjema(hentOppgaveResponseTo.getJournalpostId()) :
                hentSkjemaService.readSkjema();
        Map<String, String> answersMap = new ObjectMapper().readValue(serviceklage.getKlassifiseringJson(), new TypeReference<Map<String, String>>() {
                })
                .entrySet()
                .stream()
                .filter(entry -> entry.getValue() != null)
                .filter(entry -> !defaultValuesContainsEntry(skjemaResponse, entry))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

        addEntriesToQuestionAnswerMap(answersMap, skjemaResponse.getQuestions(), questionAnswerMap);
        return questionAnswerMap;
    }

    private void addEntriesToQuestionAnswerMap(Map<String, String> answersMap, List<Question> questions, Map<String, String> questionAnswerMap) {
        for (Question q : questions) {
            String questionId = q.getId();
            if (answersMap.containsKey(q.getId()) && !questionAnswerMap.containsKey(q.getText())) {
                String question = getQuestionById(questions, questionId)
                        .orElseThrow(() -> new ServerErrorException("Finner ikke spørsmål med id=" + questionId))
                        .getText();
                questionAnswerMap.put(question, answersMap.get(questionId));
            }

            if (q.getType().equals(Question.TypeEnum.RADIO)) {
                Optional<Answer> answer = q.getAnswers().stream()
                        .filter(a -> a.getAnswer().equals(answersMap.get(questionId)))
                        .findFirst();

                if (answer.isPresent()
                        && answer.get().getQuestions() != null
                        && !answer.get().getQuestions().isEmpty()
                        && !NONE.equals(answer.get().getNext())) {
                    addEntriesToQuestionAnswerMap(answersMap, answer.get().getQuestions(), questionAnswerMap);
                }
            }
        }
    }

    private boolean defaultValuesContainsEntry(HentSkjemaResponse skjemaResponse, Map.Entry entry) {
        if (skjemaResponse.getDefaultAnswers() == null || skjemaResponse.getDefaultAnswers().getAnswers() == null) {
            return false;
        }
        return skjemaResponse.getDefaultAnswers().getAnswers().containsKey(entry.getKey().toString());
    }

    private void opprettSlettingOppgave(HentOppgaveResponseTo hentOppgaveResponseTo) {
        OpprettOppgaveRequestTo opprettOppgaveRequestTo = opprettOppgaveRequestToMapper.mapSlettingOppgave(hentOppgaveResponseTo);
        OpprettOppgaveResponseTo opprettOppgaveResponseTo = oppgaveConsumer.opprettOppgave(opprettOppgaveRequestTo);
        log.info("Opprettet oppgave med oppgaveId={}", opprettOppgaveResponseTo.getId());
    }

    private Serviceklage getOrCreateServiceklage(String journalpostId) {
        Serviceklage serviceklage = serviceklageRepository.findByJournalpostId(journalpostId);
        if (serviceklage == null) {
            serviceklage = new Serviceklage();
            serviceklage.setJournalpostId(journalpostId);
            serviceklage.setOpprettetDato(LocalDateTime.now());
        }
        return serviceklage;
    }

    private void updateServiceklage(Serviceklage serviceklage, KlassifiserServiceklageRequest request) {
        serviceklage.setBehandlesSomServiceklage(request.getBEHANDLESSOMSERVICEKLAGE());
        serviceklage.setBehandlesSomServiceklageUtdypning(request.getBEHANDLESSOMSERVICEKLAGEUTDYPNING());
        serviceklage.setFremmetDato(request.getFREMMETDATO() == null ? null : LocalDate.parse(request.getFREMMETDATO()));
        serviceklage.setInnsender(request.getINNSENDER());
        serviceklage.setKanal(request.getKANAL());
        serviceklage.setKanalUtdypning(request.getKANALUTDYPNING());
        serviceklage.setEnhetsnummerPaaklaget(extractEnhetsnummer(request.getENHETSNUMMERPAAKLAGET()));
        serviceklage.setEnhetsnummerBehandlende(JA.equals(request.getPAAKLAGETENHETERBEHANDLENDE()) ?
                extractEnhetsnummer(request.getENHETSNUMMERPAAKLAGET()) :
                extractEnhetsnummer(request.getENHETSNUMMERBEHANDLENDE()));
        serviceklage.setGjelder(request.getGJELDER());
        serviceklage.setBeskrivelse(request.getBESKRIVELSE());
        serviceklage.setYtelse(request.getYTELSE());
        serviceklage.setRelatert(request.getRELATERT());
        serviceklage.setTema(request.getTEMA());
        serviceklage.setTemaUtdypning(mapTemaUtdypning(request));
        serviceklage.setUtfall(request.getUTFALL());
        serviceklage.setAarsak(request.getAARSAK());
        serviceklage.setTiltak(request.getTILTAK());
        serviceklage.setSvarmetode(request.getSVARMETODE());
        serviceklage.setSvarmetodeUtdypning(mapSvarmetodeUtdypning(request));
        serviceklage.setAvsluttetDato(LocalDateTime.now());
        try {
            serviceklage.setKlassifiseringJson(new ObjectMapper().writeValueAsString(request));
        } catch (JsonProcessingException e) {
            throw new ServerErrorException("Kan ikke konvertere klassifiseringsrequest til JSON-string", e);
        }
    }

    private void ferdigstillOppgave(HentOppgaveResponseTo hentOppgaveResponseTo) {
        EndreOppgaveRequestTo endreOppgaveRequestTo = endreOppgaveRequestToMapper.mapFerdigstillRequest(hentOppgaveResponseTo);
        oppgaveConsumer.endreOppgave(endreOppgaveRequestTo);
    }

    private String mapTemaUtdypning(KlassifiserServiceklageRequest request) {
        if (!isBlank(request.getVENTE())) {
            return request.getVENTE();
        } else if (!isBlank(request.getTILGJENGELIGHET())) {
            return request.getTILGJENGELIGHET();
        } else if (!isBlank(request.getINFORMASJON())) {
            return request.getINFORMASJON();
        } else if (!isBlank(request.getVEILEDNING())) {
            return request.getVEILEDNING();
        } else if (!isBlank(request.getTEMAUTDYPNING())) {
            return request.getTEMAUTDYPNING();
        } else {
            return null;
        }
    }

    private String mapSvarmetodeUtdypning(KlassifiserServiceklageRequest request) {
        if (!isBlank(request.getSVARIKKENOEDVENDIG())) {
            if (request.getSVARIKKENOEDVENDIG().equals(ANNET)) {
                return request.getSVARMETODEUTDYPNING();
            } else {
                return request.getSVARIKKENOEDVENDIG();
            }
        } else {
            return null;
        }
    }

    private String extractEnhetsnummer(String enhet) {
        if (isBlank(enhet)) {
            return null;
        }
        int l = enhet.trim().length();
        if (l >= 4) {
            String enhetsnummer = enhet.trim().substring(l - 4);
            if (isNumeric(enhetsnummer)) {
                return enhetsnummer;
            }
        }
        throw new ClientErrorException("Klarer ikke å hente ut enhetsnummer for enhet=" + enhet);
    }
}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/rest/serviceklage/service/KlassifiserServiceklageService.kt
package no.nav.tilbakemeldingsmottak.rest.serviceklage.service

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import no.nav.tilbakemeldingsmottak.bigquery.serviceklager.ServiceklageEventType
import no.nav.tilbakemeldingsmottak.bigquery.serviceklager.ServiceklagerBigQuery
import no.nav.tilbakemeldingsmottak.config.Constants.AZURE_ISSUER
import no.nav.tilbakemeldingsmottak.consumer.oppgave.OppgaveConsumer
import no.nav.tilbakemeldingsmottak.consumer.oppgave.domain.HentOppgaveResponseTo
import no.nav.tilbakemeldingsmottak.consumer.saf.SafJournalpostQueryService
import no.nav.tilbakemeldingsmottak.consumer.saf.journalpost.Journalpost
import no.nav.tilbakemeldingsmottak.domain.Serviceklage
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.ANNET
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.JA
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.KOMMUNAL_KLAGE
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.NONE
import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException
import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorNotFoundException
import no.nav.tilbakemeldingsmottak.exceptions.ErrorCode
import no.nav.tilbakemeldingsmottak.exceptions.ServerErrorException
import no.nav.tilbakemeldingsmottak.model.Answer
import no.nav.tilbakemeldingsmottak.model.KlassifiserServiceklageRequest
import no.nav.tilbakemeldingsmottak.model.Question
import no.nav.tilbakemeldingsmottak.repository.ServiceklageRepository
import no.nav.tilbakemeldingsmottak.rest.common.pdf.PdfService
import no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support.EndreOppgaveRequestToMapper
import no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support.OpprettOppgaveRequestToMapper
import no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support.ServiceklageMailHelper
import no.nav.tilbakemeldingsmottak.util.OidcUtils
import no.nav.tilbakemeldingsmottak.util.SkjemaUtils.Companion.getQuestionById
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.LocalDateTime

@Service
class KlassifiserServiceklageService(
    private val serviceklageRepository: ServiceklageRepository,
    private val oppgaveConsumer: OppgaveConsumer,
    private val endreOppgaveRequestToMapper: EndreOppgaveRequestToMapper,
    private val hentSkjemaService: HentSkjemaService,
    private val pdfService: PdfService,
    private val mailHelper: ServiceklageMailHelper,
    private val oidcUtils: OidcUtils,
    private val opprettOppgaveRequestToMapper: OpprettOppgaveRequestToMapper,
    private val serviceklagerBigQuery: ServiceklagerBigQuery,
    private val safJournalpostQueryService: SafJournalpostQueryService
) {

    private val log = LoggerFactory.getLogger(javaClass)

    @Value("\${email_serviceklage_address}")
    private lateinit var toAddress: String

    @Value("\${email_from_address}")
    private lateinit var fromAddress: String
    fun klassifiserServiceklage(request: KlassifiserServiceklageRequest, hentOppgaveResponseTo: HentOppgaveResponseTo) {
        if (KOMMUNAL_KLAGE == request.BEHANDLES_SOM_SERVICEKLAGE) {
            log.info("Klagen har blitt markert som en kommunal klage. Oppretter oppgave om sletting av dokument.")
            opprettSlettingOppgave(hentOppgaveResponseTo)
        }

        val serviceklage = getOrCreateServiceklage(hentOppgaveResponseTo.journalpostId)
        updateServiceklage(serviceklage, request)
        serviceklageRepository.save(serviceklage)
        serviceklagerBigQuery.insertServiceklage(serviceklage, ServiceklageEventType.KLASSIFISER_SERVICEKLAGE)

        log.info(
            "Serviceklage med serviceklageId={} er klassifisert som {}",
            serviceklage.serviceklageId,
            serviceklage.tema
        )
        log.info(
            "Ferdigstille oppgave med oppgaveId={} og versjonsnummer={}",
            hentOppgaveResponseTo.id,
            hentOppgaveResponseTo.versjon
        )

        ferdigstillOppgave(hentOppgaveResponseTo)
        log.info("Ferdigstilt oppgave med oppgaveId={}", hentOppgaveResponseTo.id)
        if (JA == request.KVITTERING) {
            try {
                sendKvittering(serviceklage, hentOppgaveResponseTo)
            } catch (e: Exception) {
                log.warn("Kunne ikke produsere kvittering på mail", e)
            }
        }
    }

    private fun sendKvittering(serviceklage: Serviceklage, hentOppgaveResponseTo: HentOppgaveResponseTo) {
        val email = oidcUtils.getEmailForIssuer(AZURE_ISSUER) ?: throw ClientErrorNotFoundException(
            message = "Fant ikke email-adresse i token",
            errorCode = ErrorCode.TOKEN_EMAIL_MISSING
        )

        log.info("Kvittering på innsendt klassifiseringsskjema sendes til epost: {}", email)

        val questionAnswerMap = createQuestionAnswerMap(serviceklage, hentOppgaveResponseTo)
        val pdf = pdfService.opprettKlassifiseringPdf(questionAnswerMap)
        mailHelper.sendEmail(
            fromAddress = fromAddress,
            toAddress = email,
            subject = "Kvittering på innsendt klassifiseringsskjema",
            text = "Serviceklage med oppgave-id " + hentOppgaveResponseTo.id + " har blitt klassifisert. " +
                    "Innholdet i ditt utfylte skjema ligger vedlagt.",
            fysiskDokument = pdf
        )
        log.info("Kvittering sendt på mail til saksbehandler")
    }

    private fun createQuestionAnswerMap(
        serviceklage: Serviceklage,
        hentOppgaveResponseTo: HentOppgaveResponseTo
    ): Map<String, String> {
        val questionAnswerMap = HashMap<String, String>()
        val skjemaResponse = if (hentOppgaveResponseTo.journalpostId != null) {
            hentSkjemaService.hentSkjema(hentOppgaveResponseTo.journalpostId)
        } else {
            hentSkjemaService.readSkjema()
        }
        val answersMap = serviceklage.klassifiseringJson?.let { klassifiseringsJson ->
            jacksonObjectMapper().readValue<Map<String, String?>>(klassifiseringsJson)
                .filterValues { value -> value != null }
        }

        if (answersMap != null) {
            skjemaResponse.questions?.let { addEntriesToQuestionAnswerMap(answersMap, it, questionAnswerMap) }
        }

        return questionAnswerMap
    }

    private fun addEntriesToQuestionAnswerMap(
        answersMap: Map<String, String?>,
        questions: List<Question>?,
        questionAnswerMap: MutableMap<String, String>
    ) {
        questions?.let { questionList ->
            for (question in questionList) {
                if (answersMap.containsKey(question.id) && !questionAnswerMap.containsKey(question.text)) {
                    val foundQuestion = question.id?.let { getQuestionById(questionList, it) }
                        ?: throw ServerErrorException("Finner ikke spørsmål med id=${question.id}")

                    val questionText = foundQuestion.text
                    if (questionText != null) {
                        questionAnswerMap[questionText] = answersMap[foundQuestion.id] ?: ""
                    }
                }

                if (question.type == Question.Type.RADIO) {
                    val answer = question.answers?.stream()
                        ?.filter { (answer): Answer -> answer == answersMap[question.id] }
                        ?.findFirst()
                    if (answer?.isPresent == true && answer.get().questions != null && answer.get().questions?.isNotEmpty() == true
                        && NONE != answer.get().next
                    ) {
                        addEntriesToQuestionAnswerMap(answersMap, answer.get().questions, questionAnswerMap)
                    }
                }
            }
        }
    }

    private fun opprettSlettingOppgave(hentOppgaveResponseTo: HentOppgaveResponseTo) {
        val opprettOppgaveRequestTo = opprettOppgaveRequestToMapper.mapSlettingOppgave(hentOppgaveResponseTo)
        val oppgave = oppgaveConsumer.opprettOppgave(opprettOppgaveRequestTo)
        log.info("Opprettet oppgave med oppgaveId={}", oppgave.id)
    }

    private fun getOrCreateServiceklage(journalpostId: String?): Serviceklage {
        if (journalpostId == null) {
            throw ClientErrorException("JournalpostId kan ikke være null")
        }

        var serviceklage = serviceklageRepository.findByJournalpostId(journalpostId)
        if (serviceklage == null) {
            val journalpost = getJournalPost(journalpostId)
            serviceklage = Serviceklage(
                journalpostId = journalpostId,
                opprettetDato = journalpost.datoOpprettet,
                klagenGjelderId = journalpost.bruker.id
            )
        }
        return serviceklage
    }

    private fun getJournalPost(journalpostId: String): Journalpost {
        val authorizationHeader = "Bearer " + oidcUtils.getFirstValidToken()
        return safJournalpostQueryService.hentJournalpost(journalpostId, authorizationHeader)
    }

    private fun updateServiceklage(serviceklage: Serviceklage, request: KlassifiserServiceklageRequest) {
        serviceklage.behandlesSomServiceklage = request.BEHANDLES_SOM_SERVICEKLAGE
        serviceklage.behandlesSomServiceklageUtdypning = request.BEHANDLES_SOM_SERVICEKLAGE_UTDYPNING
        serviceklage.fremmetDato = if (request.FREMMET_DATO == null) null else LocalDate.parse(request.FREMMET_DATO)
        serviceklage.innsender = request.INNSENDER
        serviceklage.kanal = request.KANAL
        serviceklage.kanalUtdypning = request.KANAL_UTDYPNING
        serviceklage.enhetsnummerPaaklaget = extractEnhetsnummer(request.ENHETSNUMMER_PAAKLAGET)
        serviceklage.enhetsnummerBehandlende =
            if (JA == request.PAAKLAGET_ENHET_ER_BEHANDLENDE) extractEnhetsnummer(request.ENHETSNUMMER_PAAKLAGET) else extractEnhetsnummer(
                request.ENHETSNUMMER_BEHANDLENDE
            )
        serviceklage.gjelder = request.GJELDER
        serviceklage.beskrivelse = request.BESKRIVELSE
        serviceklage.ytelse = request.YTELSE
        serviceklage.relatert = request.RELATERT
        serviceklage.tema = request.TEMA
        serviceklage.temaUtdypning = mapTemaUtdypning(request)
        serviceklage.utfall = request.UTFALL
        serviceklage.aarsak = request.AARSAK
        serviceklage.tiltak = request.TILTAK
        serviceklage.svarmetode = request.SVARMETODE
        serviceklage.svarmetodeUtdypning = mapSvarmetodeUtdypning(request)
        serviceklage.avsluttetDato = LocalDateTime.now()
        try {
            serviceklage.klassifiseringJson = ObjectMapper().registerKotlinModule().writeValueAsString(request)
        } catch (e: JsonProcessingException) {
            throw ServerErrorException("Kan ikke konvertere klassifiseringsrequest til JSON-string", e)
        }
    }

    private fun ferdigstillOppgave(hentOppgaveResponseTo: HentOppgaveResponseTo) {
        val endreOppgaveRequestTo = endreOppgaveRequestToMapper.mapFerdigstillRequest(hentOppgaveResponseTo)
        oppgaveConsumer.endreOppgave(endreOppgaveRequestTo)
    }

    private fun mapTemaUtdypning(request: KlassifiserServiceklageRequest): String? {
        return if (!StringUtils.isBlank(request.VENTE)) {
            request.VENTE
        } else if (!StringUtils.isBlank(request.TILGJENGELIGHET)) {
            request.TILGJENGELIGHET
        } else if (!StringUtils.isBlank(request.INFORMASJON)) {
            request.INFORMASJON
        } else if (!StringUtils.isBlank(request.VEILEDNING)) {
            request.VEILEDNING
        } else if (!StringUtils.isBlank(request.TEMA_UTDYPNING)) {
            request.TEMA_UTDYPNING
        } else {
            null
        }
    }

    private fun mapSvarmetodeUtdypning(request: KlassifiserServiceklageRequest): String? {
        return if (!StringUtils.isBlank(request.SVAR_IKKE_NOEDVENDIG)) {
            if (request.SVAR_IKKE_NOEDVENDIG == ANNET) {
                request.SVARMETODE_UTDYPNING
            } else {
                request.SVAR_IKKE_NOEDVENDIG
            }
        } else {
            null
        }
    }

    private fun extractEnhetsnummer(enhet: String?): String? {
        if (StringUtils.isBlank(enhet)) {
            return null
        }
        val l = enhet?.trim { it <= ' ' }?.length
        if (l != null) {
            if (l >= 4) {
                val enhetsnummer = enhet.trim { it <= ' ' }.substring(l - 4)
                if (StringUtils.isNumeric(enhetsnummer)) {
                    return enhetsnummer
                }
            }
        }
        throw ClientErrorException("Klarer ikke å hente ut enhetsnummer for enhet=$enhet")
    }


}
